File: Ahorn/entities/ReskinnableJellyfish.jl
module VivHelperReskinnableJelly
using ..Ahorn, Maple
using Ahorn.VivHelper

@mapdef Entity "VivHelper/ReskinnableJelly" ReskinJelly(x::Integer, y::Integer, bubble::Bool=false, tutorial::Bool=false,
Directory::String="objects/glider", Depth::Integer=-5,
GlidePath::String="particles/rect", GlideColor1::String="4FFFF3", GlideColor2::String="FFF899",
GlowPath::String="", GlowColor1::String="B7F3FF", GlowColor2::String="F4FDFF")

const placements = Ahorn.PlacementDict(
    "Reskinnable Jellyfish (Custom Particles) (Viv's Helper)" => Ahorn.EntityPlacement(
        ReskinJelly,
        "point"
    )
)

Ahorn.editingOptions(entity::ReskinJelly) = Dict{String, Any}(
    "Directory" => String["objects/glider"],
    "Depth" => merge(VivHelper.Depths, Dict{String, Integer}("Default" => -5)),
    "GlidePath" => merge(VivHelper.Particles, Dict{String, Any}("Default" => "particles/rect")),
    "GlowPath" => merge(VivHelper.Particles, Dict{String, Any}("Default" => "")),
    "GlideColor1" => VivHelper.XNAColors,
    "GlideColor2" => VivHelper.XNAColors,
    "GlowColor1" => VivHelper.XNAColors,
    "GlowColor2" => VivHelper.XNAColors,
)

function getSprite(entity::ReskinJelly)
    s = get(entity.data, "Directory", "objects/glider")
    if (endswith(s, "/")) s = chop(s); end
    sprites = Ahorn.findTextureAnimations(string(s, "/idle"), Ahorn.getAtlas());
    return sprites[1];
end

function Ahorn.selection(entity::ReskinJelly)
    x, y = Ahorn.position(entity)

    return Ahorn.getSpriteRectangle(getSprite(entity), x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::ReskinJelly, room::Maple.Room)

    Ahorn.drawSprite(ctx, getSprite(entity), 0, 0)

    if get(entity.data, "bubble", false)
        curve = Ahorn.SimpleCurve((-7, -1), (7, -1), (0, -6))
        Ahorn.drawSimpleCurve(ctx, curve, (1.0, 1.0, 1.0, 1.0), thickness=1)
    end
end

end

File: Loenn/entities/ReskinnableJellyfish.lua
local utils = require('utils')
local vivUtil = require('mods').requireFromPlugin('libraries.vivUtil')
local drawableSprite = require("structs.drawable_sprite")
local drawableLine = require("structs.drawable_line")
local drawing = require("utils.drawing")

local reskinJelly = {
    name = "VivHelper/ReskinnableJellyfish",
    fieldOrder = {
        "x","y",
        'bubble','tutorial',
        'Directory','Depth',
        'GlidePath','GlowPath',
        'GlideColor1','GlideColor2',
        'GlowColor1','GlowColor2'        
    },
    fieldInformation = {
        Directory = vivUtil.getDirectoryPathFromFile(false),
        Depth = {fieldType = 'integer'},
        GlidePath = vivUtil.GetFilePathWithNoTrailingNumbers(true),
        GlowPath = vivUtil.GetFilePathWithNoTrailingNumbers(true),
        GlideColor1 = {fieldType = 'color', allowXNAColors = true, allowEmpty = true },
        GlideColor2 = {fieldType = 'color', allowXNAColors = true, allowEmpty = true },
        GlowColor1 = {fieldType = 'color', allowXNAColors = true, allowEmpty = true },
        GlowColor2 = {fieldType = 'color', allowXNAColors = true, allowEmpty = true }
    }
}

reskinJelly.placements = {
    name = "jelly",
    data = {
        bubble = false, tutorial = false,
        Directory = "objects/glider", Depth = -5,
        GlidePath = "particles/rect", GlowPath = "",
        GlideColor1="4FFFF3", GlideColor2="FFF899",
        GlowColor1="B7F3FF", GlowColor2="F4FDFF"
    }
}

reskinJelly.sprite = function(room,entity)
    local sprite = vivUtil.getImageWithNumbers(entity.Directory .. "/idle", 0, entity)
    if entity.bubble then
        local x, y = entity.x or 0, entity.y or 0
        local points = drawing.getSimpleCurve({x - 11, y - 1}, {x + 11, y - 1}, {x - 0, y - 6})
        local lineSprites = drawableLine.fromPoints(points):getDrawableSprite()

        table.insert(lineSprites, sprite)

        return lineSprites
    else
        return sprite 
    end
end

return reskinJelly